README
IEnabler: A PyGtk+ IEnabler client for the University of Canterbury


debian/changelog
ienabler (1.0-1ubuntu2) jaunty; urgency=low

  * Rebuild

 -- John Stowers <john.stowers@gmail.com>  Sat, 01 Aug 2009 19:58:10 +1200

ienabler (1.0-1ubuntu1) jaunty; urgency=low

  * Rebuild

 -- John Stowers <john.stowers@gmail.com>  Sat, 01 Aug 2009 19:03:53 +1200

ienabler (1.0-1) jaunty; urgency=low

  * Initial release

 -- John Stowers <john.stowers@gmail.com>  Sat, 01 Aug 2009 17:24:42 +1200


ienabler.py
#!/usr/bin/env python
# GUI for logging into the University of Canterbury internet
# Works like ienabler for the windows machines
#
# (c) John Stowers 2008

import telnetlib

class IEnabler(object):
    def __init__(self,user,password, host="ienabler.canterbury.ac.nz", port=259):
        self.user = user
        self.password = password
        try:
            self.tn = telnetlib.Telnet(
                            host,
                            port
                            )
        except Exception, e:
            self.tn = None
            print "ERROR: %s" % e

    def _read_string(self, string):
        result = self.tn.read_until(string, 5)
        return result.endswith(string)

    def _login(self):
        if self._read_string("User: "):
            self.tn.write(self.user + "\n")
            if self._read_string("password: "):
                self.tn.write(self.password + "\n")
                return True
        return False

    def _choice(self,choice):
        try:
            if self._login():
                #returns the idx of the matched regex
                idx,match,txt = self.tn.expect(
                                    [r'.*Access denied.*',r'.*Enter your choice.*'],
                                    5
                                    )
                if idx == 1:
                    self.tn.write("%s\n" % choice)
                    if self._read_string('\n'):
                        return True
                else:
                    print "ERROR: %s" % txt
                    return False
        except EOFError:
            return False

    def enable(self):
        if self.tn:
            ok = self._choice(1)
            self.tn.close()
        else:
            ok = False
        return ok

    def disable(self):
        if self.tn:
            ok = self._choice(2)
            self.tn.close()
        else:
            ok = False
        return ok




bdist_debian.py
# bdist_debian.py
#
# Add 'bdist_debian' Debian binary package distribution support to 'distutils'.
#
# Written by: Gene Cash <gene.cash@gmail.com> 16-NOV-2007
# Extended by: Chris Freeze <cfreeze@users.sourceforge.net> 11/08/2008
#       * Removed Nokia specific items
#       * Added other keywords that can be in the control file

import os, base64
from distutils.core import Command, Distribution
from distutils.dir_util import remove_tree
from distutils.util import byte_compile

# make these legal keywords for setup()
Distribution.icon=None
Distribution.section=None
Distribution.depends=None
Distribution.recommends=None
Distribution.suggests=None
Distribution.architecture=None
Distribution.essential=None
Distribution.source=None

class ControlFile(object):
    def __init__(self, Installed_Size=0, Long_Description='', Description='', Icon='', **kwargs):
        self.options=kwargs
        self.description=Description
        self.long_description=Long_Description
        self.icon=Icon
        self.installed_size=Installed_Size

    def getContent(self):
        content=['%s: %s' % (k, v) for k,v in self.options.iteritems()]

        content.append('Installed-Size: %d' % self.installed_size)
        if self.description != 'UNKNOWN':
            content.append('Description: %s' % self.description.strip())
            if self.long_description != 'UNKNOWN':
                self.long_description=self.long_description.replace('\n', '\n ')
                content.append(' '+self.long_description.strip())

        if self.icon:
            # generate Base64-encoded icon
            s=file(self.icon, 'rb').read()
            x=base64.b64encode(s)
            # wrap width MUST be 76 characters to make application manager happy after install
            lw=76
            # trailing blank is important, and the XB- is NOT legal
            content.append('XB-Maemo-Icon-26: ')
            for i in range(0, len(x), lw):
                content.append(' '+x[i:i+lw])

        # must have two returns
        return '\n'.join(content)+'\n\n'

class bdist_debian(Command):
    description=''
    # List of option tuples: long name, short name (None if no short name), and help string.
    user_options=[('name=', None, 'Package name'),
                  ('section=', None, 'Section (Only "user/*" will display in App Mgr usually) (default=user/other)'),
                  ('priority=', None, 'Priority  (default=optional)'),
                  ('depends=', None, 'Other Debian package dependencies (comma separated) (default=python)'),
                  ('recommends=', None, 'Other Debian package you recommend (comma separated) (default=none)'),
                  ('suggests=', None, 'Other Debian package you suggest (comma separated) (default=none)'),
                  ('essential=', None, 'Whether packge is essential(yes/no) (default=no)'),
                  ('architecture=', None, 'Target architecture (default=all)'),
                  ('source=', None, 'Source keyword.. (default=package_name)'),
                  ('icon=', None, 'Name of icon file to be displayed by App Mgr')]

    def initialize_options(self):
        self.section=None
        self.priority=None
        self.depends=None
        self.recommends=None
        self.suggests=None
        self.architecture=None
        self.essential=None
        self.source=None
        self.icon=None

    def finalize_options(self):
        if self.section is None:
            self.section='user/other'

        if self.priority is None:
            self.priority='optional'

        self.maintainer='%s <%s>' % (self.distribution.get_maintainer(), self.distribution.get_maintainer_email())

        if self.depends is None:
            self.depends='python'

        self.name=self.distribution.get_name()
        self.description=self.distribution.get_description()
        self.long_description=self.distribution.get_long_description()
        self.version=self.distribution.get_version()

        if self.source is None:
            self.source=self.name

        # process new keywords
        if self.distribution.icon != None:
            self.icon=self.distribution.icon
        if self.distribution.section != None:
            self.section=self.distribution.section
        if self.distribution.depends != None:
            self.depends=self.distribution.depends
        if self.distribution.recommends != None:
            self.recommends=self.distribution.recommends
        if self.distribution.suggests != None:
            self.suggests=self.distribution.suggests
        if self.distribution.architecture != None:
            self.architecture=self.distribution.architecture
        if self.distribution.essential != None:
            self.essential=self.distribution.essential

    def run(self):
        build_dir = os.path.join(self.get_finalized_command('build').build_base, 'debian')
        dist_dir = 'dist'
        binary_fn = 'debian-binary'
        control_fn = 'control'
        data_fn = 'data'
        tgz_ext = '.tar.gz'

        # build everything locally
        self.run_command('build')
        install=self.reinitialize_command('install', reinit_subcommands=1)
        install.root=build_dir
        self.run_command('install')

        # find out how much space it takes
        installed_size=0
        for root, dirs, files in os.walk('build'):
            installed_size+=sum(os.path.getsize(os.path.join(root, name)) for name in files)

        # make compressed tarball
        self.make_archive(os.path.join(dist_dir, data_fn), 'gztar', root_dir=build_dir)

        # remove all the stuff we just built
        remove_tree(build_dir)

        # create control file contents
        ctl = ControlFile(
            Package = self.name, 
            Version = self.version, 
            Section = self.section, 
            Priority = self.priority,
            Installed_Size = (installed_size / 1024) + 1, 
            Architecture = self.architecture, 
            Maintainer = self.maintainer,
            Depends = self.depends, 
            Description = self.description, 
            Long_Description = self.long_description,
            Recommends = self.recommends, 
            Suggests = self.suggests, 
            Essential = self.essential,
			Source = self.source,
            Icon=self.icon).getContent()

        # grab scripts
        scripts={}
        for fn in ('postinst', 'preinst', 'postrm', 'prerm', 'config'):
            if os.path.exists(fn):
                scripts[fn]=file(fn, 'rb').read()

        # now to create the deb package
        os.chdir(dist_dir)

        # write control file
        file(control_fn, 'wb').write(ctl)

        # write any scripts and chmod a+rx them
        files=[control_fn]
        for fn in scripts:
            files.append(fn)
            file(fn, 'wb').write(scripts[fn])
            os.chmod(fn, 0555)

        # make "control" compressed tarball with control file and any scripts
        self.spawn(['tar', '-czf', control_fn+tgz_ext]+files)

        # make debian-binary file
        file(binary_fn, 'wb').write('2.0\n')

        # make final archive
        package_filename='%s_%s_all.deb' % (self.name, self.version)
        if 1:
            self.spawn(['ar', '-cr', package_filename, binary_fn, control_fn+tgz_ext, data_fn+tgz_ext])

		    # Remove unneeded files
            os.remove(control_fn)
            os.remove(control_fn+tgz_ext)
            os.remove(data_fn+tgz_ext)
            os.remove(binary_fn)
        else:
            #fixed deb builder script
            fix_script = """#!/bin/sh
            echo Started
            mkdir -p tmp/DEBIAN
            cd tmp
            tar -xzvpf ../data.tar.gz
            cd DEBIAN
            tar -xzvpf ../../control.tar.gz
            cp ../../debian-binary ./
            # Make some tweak to preinst/postinst/etc...
            cd ../../
            dpkg-deb -b tmp %s
            # new.deb should not cause the same complaints
            rm -r tmp
            """ % (package_filename)
            fix_deb = "fix_deb"
            
            #write the script & execute it
            f = file(fix_deb, 'w')
            f.write(fix_script)
            f.close()

            os.chmod(fix_deb, 0755)
            os.system('./%s' % fix_deb)
            os.remove(control_fn)
            os.remove(control_fn+tgz_ext)
            os.remove(data_fn+tgz_ext)
            os.remove(binary_fn)
            os.remove(fix_deb)


ienabler-cmd.py
#!/usr/bin/env python
# Script for logging into the University of Canterbury internet
# Works like ienabler for the windows machines
#
# (c) John Stowers 2008

import sys
import getopt
import getpass
from ienabler import IEnabler

if __name__ == "__main__":
    #FILL IN DEFAULTS IF YOU WANT
    USER = ""
    PASSWORD = ""
    HOST = "ienabler.canterbury.ac.nz"
    PORT = "259"

    try:
        opts, args = getopt.getopt(sys.argv[1:], "du:p:h:P:", ["disable", "user", "password", "host", "port"])
    except getopt.GetoptError:
        print   "Ienabler: Enables and Disables UOC Internet Access.\n" + \
                "Usage:\n" + \
                " ienabler [OPTIONS]\n" + \
                "Options:\n" + \
                " -d,--disable   Logs you out of the internet\n" + \
                " -u,--user      Username\n" + \
                " -p,--pasword   Password\n" + \
                " -h,--host      ienabler host\n" + \
                " -P,--port      ienabler port\n"
        sys.exit(1)

    choice = "Enable"
    for o, a in opts:
        if o in ("-d", "--disable"):
            choice = "Disable"
        if o in ("-u", "--user"):
            USER = a
        if o in ("-p", "--password"):
            PASSWORD = a
        if o in ("-h", "--host"):
            HOST = a
        if o in ("-P", "--port"):
            PORT = a


    if USER == "":
        USER = raw_input("Username:")
    if PASSWORD == "":
        PASSWORD = getpass.getpass("Password:")

    i = IEnabler(USER,PASSWORD,HOST,PORT)
    if choice == "Enable":
        ok = i.enable()
    else:
        ok = i.disable()

    if ok:
        print "%sd OK" % choice
    else:
        print "Could not %s" % choice
        sys.exit(1)





ienabler-gui.py
#!/usr/bin/env python
# GUI for logging into the University of Canterbury internet
# Works like ienabler for the windows machines
#
# (c) John Stowers 2008

import gtk
import gobject
import dbus.glib
import pynotify
import threading
import os.path
import webbrowser
import ConfigParser
from ienabler import IEnabler

CONFIGURATION_DEFAULTS = {
    "name"              :       "IEnabler",
    "user"              :       "",
    "password"          :       "",
    "delay_ms"          :       "500", #< 0 disables auto login at start
    "add_funds_url"     :       "https://ucstudentweb.canterbury.ac.nz/",
    "ienabler_host"     :       "ienabler.canterbury.ac.nz", #123.255.13.254
    "ienabler_port"     :       "259",
}
CONFIGURATION_NAMES = (
    ("user","Username: "),
    ("password","Password: ")
)

class Config:
    def __init__(self):
        self._filePath = os.path.join(os.environ["HOME"],".ienabler")
        self._config = ConfigParser.ConfigParser(defaults=CONFIGURATION_DEFAULTS)
        self._config.read(self._filePath)

    def get(self, key):
        return self._config.get('DEFAULT',key)

    def set(self, key, value):
        self._config.set('DEFAULT',key, value)

    def save(self):
        fp = open(self._filePath, 'w')
        self._config.write(fp)
        fp.close()
CONFIGURATION = Config()

class Authenticator(threading.Thread, gobject.GObject):
    __gsignals__ =  { 
                    "completed": (gobject.SIGNAL_RUN_LAST, gobject.TYPE_NONE, [gobject.TYPE_BOOLEAN,gobject.TYPE_STRING])
                    }

    def __init__(self, choice):
        gobject.GObject.__init__(self)
        threading.Thread.__init__(self)
        self.choice = choice

    def run(self):
        i = IEnabler(
                user=CONFIGURATION.get("user"),
                password=CONFIGURATION.get("password"),
                host=CONFIGURATION.get("ienabler_host"),
                port=CONFIGURATION.get("ienabler_port")
                )
        if self.choice == "Enable":
            ok = i.enable()
        else:
            ok = i.disable()
        self.emit("completed", ok, self.choice)

    def emit(self, *args):
        """
        Override the gobject signal emission so that signals
        can be emitted from threads
        """
        gobject.idle_add(gobject.GObject.emit,self,*args)

class NetworkListener(gobject.GObject):

    SERVICE_NAME = "org.freedesktop.NetworkManager"
    SERVICE_PATH = "/org/freedesktop/NetworkManager"

    __gsignals__ =  { 
                "online": (gobject.SIGNAL_RUN_LAST, gobject.TYPE_NONE, []),
                "offline": (gobject.SIGNAL_RUN_LAST, gobject.TYPE_NONE, [])
                }

    def __init__(self):
        gobject.GObject.__init__(self)
        self.online = False
        try:
                systemBus = dbus.SystemBus()
                proxy_obj = systemBus.get_object(NetworkListener.SERVICE_NAME,
                                                 NetworkListener.SERVICE_PATH)
                nm_interface = dbus.Interface(proxy_obj, NetworkListener.SERVICE_NAME)

                #magic number: aparently 3 == online
                self.online = int(nm_interface.state()) == 3
                
                nm_interface.connect_to_signal('DeviceNowActive', self.active_cb)
                nm_interface.connect_to_signal('DeviceNoLongerActive', self.inactive_cb)
        except dbus.DBusException, de:
                print "Error while connecting to NetworkManager: %s" % str(de)
                self.online = True

    def active_cb(self, path):
        self.online = True
        print "ONLINE"
        self.emit("online")

    def inactive_cb(self, path):
        print "OFFLINE"
        self.online = False
        self.emit("offline")


class Gui:
    def __init__(self):
        pynotify.init(CONFIGURATION.get("name"))
        self.notifications_show_actions = 'actions' in pynotify.get_server_caps()
        self.online = False

        self._create_gui()
        self.nm = NetworkListener()
        self.nm.connect("online", lambda x: self.authenticate("Enable"))
        if self.nm.online:
            delay_ms = int(CONFIGURATION.get("delay_ms"))
            if delay_ms >= 0:
                gobject.timeout_add(delay_ms,self.authenticate,"Enable")

    def _create_gui(self):
        #load themed or fallback app icon
        try:
            icon = os.path.join(os.path.dirname(__file__),'uclogo.svg')
            if os.path.exists(icon):
                #not installed
                self.icon = gtk.gdk.pixbuf_new_from_file(icon)
            else:
                #installed
                self.icon = gtk.icon_theme_get_default().load_icon("uclogo", 24, gtk.ICON_LOOKUP_FORCE_SVG)
        except gobject.GError:
            self.icon = gtk.icon_theme_get_default().load_icon(gtk.STOCK_NETWORK, 24, gtk.ICON_LOOKUP_FORCE_SVG)

        #build tray icon
        self.tray = gtk.StatusIcon()
        self.tray.set_from_pixbuf(self.icon)
        self.tray.set_tooltip('IEnabler Connecting..')
        self.tray.connect('popup-menu', self._on_popup_menu)
        self.tray.set_visible(True)

        #attach the libnotification bubble to the tray
        self.notification = pynotify.Notification(CONFIGURATION.get("name"))
        self.notification.attach_to_status_icon(self.tray)
        self.notification.set_timeout(pynotify.EXPIRES_DEFAULT)
        
        #create popup menu
        self.menu = gtk.Menu()
        enable = gtk.ImageMenuItem(stock_id=gtk.STOCK_YES, accel_group=None)
        enable.connect("activate", lambda x: self.authenticate("Enable"))
        disable = gtk.ImageMenuItem(stock_id=gtk.STOCK_NO, accel_group=None)
        disable.connect("activate", lambda x: self.authenticate("Disable"))
        configure = gtk.ImageMenuItem(stock_id=gtk.STOCK_PREFERENCES, accel_group=None)
        configure.connect("activate", self._on_configure_clicked)
        about = gtk.ImageMenuItem(stock_id=gtk.STOCK_ABOUT, accel_group=None)
        about.connect("activate", self._on_about_clicked)
        quit = gtk.ImageMenuItem(stock_id=gtk.STOCK_QUIT, accel_group=None)
        quit.connect("activate", self._on_exit_clicked)
        
        self.menu.add(enable)
        self.menu.add(disable)
        self.menu.add(gtk.SeparatorMenuItem())
        self.menu.add(configure)
        self.menu.add(gtk.SeparatorMenuItem())
        self.menu.add(about)
        self.menu.add(quit)
        self.menu.show_all()

    def _on_popup_menu(self, status, button, time):
        self.menu.popup(None, None, gtk.status_icon_position_menu, button, time, self.tray)

    def _on_configure_clicked(self, widget):
        def make_label_and_entry(defaultValue, displayName, packInto, sizegroup):
            #makes a gtk.HBox with a label and a gtk.Entry. packs the
            #hbox into the container, Returns the entry
            lbl = gtk.Label(displayName)
            lbl.set_alignment(0, 0.5)
            sizegroup.add_widget(lbl)

            w = gtk.Entry()
            w.set_text(defaultValue)

            hb = gtk.HBox()
            hb.pack_start(lbl)
            hb.pack_start(w)

            packInto.pack_start(hb)
            return w

        #Automatically build a yuck gui for all the config options
        dlg = gtk.Dialog(
                        title="Internet Configuration",
                        buttons=(
                            gtk.STOCK_CANCEL,gtk.RESPONSE_REJECT,
                            gtk.STOCK_OK,gtk.RESPONSE_ACCEPT))
        #make all text entries the same size
        labelSizeGroup = gtk.SizeGroup(gtk.SIZE_GROUP_HORIZONTAL)

        #save widgets so we can get values from them later
        widgets = {}
        for pref,display in CONFIGURATION_NAMES:
            widgets[pref] = make_label_and_entry(
                                defaultValue=CONFIGURATION.get(pref),
                                displayName=display,
                                packInto=dlg.vbox,
                                sizegroup=labelSizeGroup)

        #hide password text
        widgets["password"].set_visibility(False)

        dlg.show_all()
        resp = dlg.run()
        if resp == gtk.RESPONSE_ACCEPT:
            for pref,display in CONFIGURATION_NAMES:
                w = widgets[pref]
                CONFIGURATION.set(pref, w.get_text())
        dlg.destroy()

    def _on_about_clicked(self, widget):
        about = gtk.AboutDialog()
        about.set_name(CONFIGURATION.get("name"))
        about.set_logo(self.icon)
        about.set_authors(("John Stowers",))
        about.set_comments("Log to the University of Canterbury Internet")
        about.run()
        about.destroy()

    def _on_exit_clicked(self, widget):
        if self.is_online():
            self.authenticate("Disable", block=True)
        CONFIGURATION.save()
        gtk.main_quit()

    def _on_add_funds(self, n, action):
        webbrowser.open(CONFIGURATION.get("add_funds_url"))
        n.close()

    def _on_authenticated(self, authenticator, ok, choice):
        self.notification.clear_actions()
        if ok:
            msg = "Internet Access %sd OK" % choice
            self.online = choice == "Enable"
        else:
            msg = "Could not %s Internet Access" % choice
            self.online = False
            if choice == "Enable" and self.notifications_show_actions:
                self.notification.add_action("add_funds", "Add Funds", self._on_add_funds)
        self.tray.set_tooltip(msg)
        self.notification.update(CONFIGURATION.get("name"),msg,gtk.STOCK_NETWORK)
        self.notification.show()
                
    def authenticate(self, choice, block=False):
        authenticator = Authenticator(choice)
        authenticator.connect("completed", self._on_authenticated)
        authenticator.start()
        if block:
            authenticator.join()

    def is_online(self):
        return self.nm.online and self.online

if __name__ == "__main__":
    gtk.gdk.threads_init()
    app = Gui()
    gtk.main()



